### ISSUE TITLE
Suggestion yii\db\Connection master/slave when master only one fail.please no cache.

### ISSUE THREAD

**Issue by netyum** [yiisoft/yii2#11306]
### What steps will reproduce the problem?

When network unstable or  network jitter . Connection serverStatusCache 
When Connection masters only one or other fail in serverStatusCache， 
Ok  In serverRetryInterval second. master not been available. but tuning serverRetryInterval no good for slave ...

```php
'db' => [
            'class' => 'yii\db\Connection',
            'enableSchemaCache' => true,
            'schemaCacheDuration' => 300,
            'schemaCache' => 'db schema cache',
            'charset' => 'utf8',
            'tablePrefix'=>'t_',
            'masterConfig' => [
                'username' => 'xxxxx',
                'password' => 'xxxxx',
                'attributes' => [
                    PDO::ATTR_TIMEOUT => 10,
                ]
            ],
            'masters' => [
                ['dsn' => 'mysql:host=db01;dbname=db'], //only one please no cache
                ['dsn' => 'mysql:host=db02;dbname=db'], //fail
                ['dsn' => 'mysql:host=db03;dbname=db'] //fail
            ],
           ....
```
### What is the expected result?

Exception message "None of the master DB servers is available." inexplicable
### What do you get instead?

Please no cache only one master.
### Additional info

| Q | A |
| --- | --- |
| Yii version |  |
| PHP version |  |
| Operating system |  |

**Comment by netyum** [comment]
```
'db' => [
            'class' => 'yii\db\Connection',
            'enableSchemaCache' => true,
            'schemaCacheDuration' => 300,
            'schemaCache' => 'db schema cache',
            'charset' => 'utf8',
            'tablePrefix'=>'t_',
            'masterConfig' => [
                'username' => 'xxxxx',
                'password' => 'xxxxx',
                'attributes' => [
                    PDO::ATTR_TIMEOUT => 10,
                ]
            ],
            'masters' => [
                ['dsn' => 'mysql:host=db01;dbname=db']
            ],
           ....
```

mysqld stop

request web  or run yii script

Ok throw execption

mysqld start

continue request web or run yii script

error still

because db server fail status was cache.

when network jitter in Server cluster. Should not last for so long without the connection faild DB.

**Comment by netyum** [comment]
```php
<?php

namespace common\components;

use PDO;
use Yii;
use yii\base\InvalidConfigException;
use yii\caching\Cache;
use yii\db\Exception;

class Connection extends \yii\db\Connection
{
    public function open()
    {
        if ($this->pdo !== null) {
            return;
        }

        if (!empty($this->masters)) {
            $db = $this->openFromPool($this->masters, $this->masterConfig, true);
            if ($db !== null) {
                $this->pdo = $db->pdo;
                return;
            } else {
                throw new InvalidConfigException('None of the master DB servers is available.');
            }
        }

        if (empty($this->dsn)) {
            throw new InvalidConfigException('Connection::dsn cannot be empty.');
        }
        $token = 'Opening DB connection: ' . $this->dsn;
        try {
            Yii::info($token, __METHOD__);
            Yii::beginProfile($token, __METHOD__);
            $this->pdo = $this->createPdoInstance();
            $this->initConnection();
            Yii::endProfile($token, __METHOD__);
        } catch (\PDOException $e) {
            Yii::endProfile($token, __METHOD__);
            throw new Exception($e->getMessage(), $e->errorInfo, (int) $e->getCode(), $e);
        }
    }

    protected function openFromPool(array $pool, array $sharedConfig, $isMaster = false)
    {
        if (empty($pool)) {
            return null;
        }

        if (!isset($sharedConfig['class'])) {
            $sharedConfig['class'] = get_class($this);
        }

        $cache = is_string($this->serverStatusCache) ? Yii::$app->get($this->serverStatusCache, false) : $this->serverStatusCache;

        //过滤可用的db
        foreach ($pool as $index => $config) {
            $key = [__METHOD__, $config['dsn'], $isMaster];
            if ($cache instanceof Cache && $cache->get($key)) {
                // should not try this dead server now
                unset($pool[$index]);
            }
        }

        shuffle($pool);


        foreach ($pool as $config) {
            $config = array_merge($sharedConfig, $config);
            if (empty($config['dsn'])) {
                throw new InvalidConfigException('The "dsn" option must be specified.');
            }
            $key = [__METHOD__, $config['dsn'], $isMaster];
            /* @var $db Connection */
            $db = Yii::createObject($config);

            try {
                $db->open();
                return $db;
            } catch (\Exception $e) {
                Yii::warning("Connection ({$config['dsn']}) failed: " . $e->getMessage(), __METHOD__);
                if ($isMaster == true && count($pool) == 1) {
                    throw new Exception($e->getMessage(), $e->errorInfo, (int) $e->getCode(), $e);
                } else {
                    if ($cache instanceof Cache) {
                        // mark this server as dead and only retry it after the specified interval
                        $cache->set($key, 1, $this->serverRetryInterval);
                    }
                }
            }
        }

        return null;
    }
}
```

**Comment by 23c** [comment]
thanks a lot @netyum

**Comment by samdark** [comment]
Re-read several times and check current code. Wasn't able to get what's the issue is about and how to reproduce it. Please comment if you're willing to elaborate it. Thanks.

**Comment by php-cpm** [comment]
Do not know why you cannot reproduce it.
mysql Master node CANNOT be removed when Cache on even when NETWORK down. That is the problem.
Default setting `serverRetryInterval` is too long.
Yii framework is too bad on handling this.

**Comment by php-cpm** [comment]
rethink this condition, if your db server is down for 5s,nyou need to wait another 10min to recover.

**Comment by php-cpm** [comment]
my now temp solution is change `serverRetryInterval` to 60, make the error time in just 1 minute.

**Comment by cebe** [comment]
@php-cpm @netyum @23c sorry for the handling of this issue, it seems we did not understand what it was about. I have reopened a ticket with clear description and we'll try to fix it soon: #17991
